Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The default look group used to schedule async work
+    let loopGroup: EventLoopGroup
+
+    /// The configuration used to bootstrap new HTTP connections
+    let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> ProviderGetAction {
+        return self.lock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .useExisting(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                self._connectionProviders[key] = .future(promise.futureResult)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
```

Comment:
is the synchronisation correct here? We take the lock to look stuff up, then we release it (by returning) and in the `promise.futureResult.whenComplete` callback, we do `self[key]` which I assume takes the lock again. But we can't make assumptions about the state anymore because we released the lock.

Reply:
> But we can't make assumptions about the state anymore because we released the lock

I think the only thing that must stay true is that the dictionary of providers (accessed through self[someKey] under the `lock`) shouldn't be modified during the `getAction` critical section, but outside of this I can't find what would cause an issue.

Reply:
What about the following:
 - thread A requests a connection with `key1`
 - we return `.make(promise1)`
 - thread B requests a connection with `key1`
 - we return another `.make(promise2)`
 - `promise1` succeeds, updating the `key1`
 - `promise2` succeeds, rewriting the `key1` with another connection
Am I missing something here? I this is indeed the case, then we can leak connections due to a missing cleanup, or we may fail to honour maxconnections per host for example. @weissi is this scenario that you had in mind?

Reply:
Ah, I see, in this case we will return `.useExisting` which is a future of the provider!

Reply:
Here is a thought/question. Why do we need `promise` for the provider? why not to create an empty provider with no initial connection, add it to the pool, and then call `getConnection`, then you don't need any futures, you don't need duplication of client bootrstrap. I've tried removing the `initialConnection` from http1 provider along with `.future` provider enum case and it seems to work fine, at least all tests pass. Wdyt?

Reply:
@artemredkin Thank you! It seems like this would be way better, and I think this can also probably help to better reason about some synchronisation issues I still have, I'm going to try this!

Reply:
@artemredkin Actually, I think remember why we went with this approach originally: there was the additional case `.http2` provider case that I was also working on, and I then dropped it, but before this we couldn't synchronously know if it was going to be an `http1` provider or an `http2` provider.

I still think its better to go with your approach and reconsider the design when we actually want to add `HTTP/2` support.